_posts/2022-12-19-running-outputs-outside-zonejs-for-angular-performance-optimization.md

```diff
@@ -0,0 +1,288 @@
+    // Subscribes to the source observable in the outer zone
+    .pipe(outsideAngular());
+
+@Output() public readonly move = fromEvent(this.document, 'mousemove')
+    .pipe(
+        outsideAngular(), // Subscribes to the source observable
+                          // in the outer zone
+        map((e: MouseEvent) => {
+            return {
+                x: e.x,
+                y: e.y,
+            };
+        })
+    );
+```
+
+Creating a custom operator in RxJS is quite straightforward. It's a function
+that returns a function that takes the source observable as a parameter.
+
+```typescript
+export const outsideAngular = <T>() => (source$: Observable<T>) => {
+    // Todo: create yourNewObservable$
+    return yourNewObservable$;
+};
+```
+
+In our case, we want to subscribe on `source$` inside the **outer zone** of Angular and keep the results of that subscription in a subject that we can return. 
+For that, we need to get access to an instance of `NgZone`... We can use the `inject()` function of Angular to get access to that instance.
+
+```typescript
+export const outsideAngular = <T>() => (source$: Observable<T>) => {
+    // Get access to ngZone
+    const ngZone = inject(NgZone);
+
+    // Create a subject of the same generic type
+    // as the source observable
+    const sub$$ = new Subject<T>();
+    
+    // Declare the subscription here, we need to clean it up later
+    let subscription: Subscription;
+
+    // Run the actual subscription that executes an
+    // addEventListener outside of Angular, so that
+    // Change Detection won't be triggered automatically
+    ngZone.runOutsideAngular(() => {
+        subscription = source$.subscribe(sub$$);
+    });
+
+    // Return the subject but be sure to unsubscribe from
+    // the source observable when the subject completes
+    // or errors
+    return sub$$.pipe(
+        finalize(() => {
+            subscription.unsubscribe();
+        })
+    );
+};
+```
+
+Everything works as expected now: `ChildComponent` does not trigger Change Detection anymore but `ParentComponent` is notified with the new values.
+
+We can see that in this [StackBlitz example](https://stackblitz.com/edit/angular-ivy-komr7r){:target="_blank"}
+The DOM is not rerendered, but `console.log` statements are being made every second and when the user moves the mouse.
+
+## Custom Change Detection
+
+`ChildComponent` does what it needs to do:
+- Run event listeners in the outer zone
+- Notify the parent through `@Output()`s without triggering Change Detection
+
+`ParentComponent` is now responsible of triggering Change Detection and we don't want to trigger it all the time. We only want to trigger it when the mouse of the user moves within the green box.
+
+
+```typescript
+export class ParentComponent {
+    // Inject ChangeDetectorRef because we want to
+    // manually trigger Change Detection with the
+    // detectChanges() function
+    private readonly cdRef= inject(ChangeDetectorRef)
```

Comment:
```suggestion
    private readonly chnageDetectorRef= inject(ChangeDetectorRef)
```

Reply:
Motivate your readers to not use abbreviations.

I personally go with `changeDetector`

Reply:
done